File: server/apps/nexotype/schemas/engineering/construct_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class ConstructCreate(BaseModel):
    """Schema for creating a new construct"""
    candidate_id: int
    plasmid_map_url: str | None = None


class ConstructUpdate(BaseModel):
    """Schema for updating a construct"""
    candidate_id: int = None
    plasmid_map_url: str | None = None


# ==========================================
# Response Schemas
# ==========================================

class ConstructDetail(BaseModel):
    """Schema for construct details"""
    id: int
    candidate_id: int
    plasmid_map_url: str | None = None
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class ConstructListResponse(BaseModel):
    """Response schema for listing constructs"""
    success: bool
    data: list[ConstructDetail] | None = None
    count: int = 0
    error: str | None = None


class ConstructResponse(BaseModel):
    """Response schema for single construct operations"""
    success: bool
    data: ConstructDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None

File: client/src/modules/nexotype/schemas/engineering/construct.schemas.ts
import { z } from 'zod';

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const ConstructSchema = z.object({
  id: z.number(),
  candidate_id: z.number().int(),
  plasmid_map_url: z.string().max(255).nullable().optional(),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateConstructSchema = z.object({
  candidate_id: z.number().int(),
  plasmid_map_url: z.string().max(255).optional(),
});

export const UpdateConstructSchema = z.object({
  candidate_id: z.number().int().optional(),
  plasmid_map_url: z.string().max(255).nullable().optional(),
});

// ==========================================
// Types
// ==========================================

export type Construct = z.infer<typeof ConstructSchema>;
export type CreateConstruct = z.infer<typeof CreateConstructSchema>;
export type UpdateConstruct = z.infer<typeof UpdateConstructSchema>;

// ==========================================
// Response Types
// ==========================================

export type ConstructResponse = {
  success: boolean;
  data?: Construct;
  error?: string;
};

export type ConstructsResponse = {
  success: boolean;
  data?: Construct[];
  count?: number;
  error?: string;
};
